File: runtime/src/test/py/ref_page.py
#
# microtome

from microtome.page import Page
from microtome.core.prop import Prop, PropSpec
from microtome.core.page_ref import PageRef
from primitive_page import PrimitivePage

class RefPage(Page):
    def __init__(self, name):
        super(RefPage, self).__init__(name)
        self._nested = Prop(self, RefPage._s_nestedSpec)

    @property
    def nested(self):
        return self._nested.value.page if self._nested.value is not None else None

    @property
    def props(self):
        return super(RefPage, self).props + [self._nested]

    _s_nestedSpec = PropSpec("nested", None, [PageRef, PrimitivePage, ])

File: runtimes/objc/Tests/RefPage.m

#import "RefPage.h"
#import "MTPageRef.h"
#import "PrimitivePage.h"
#import "microtome.h"

static MTPropSpec* s_nestedSpec = nil;

@implementation RefPage {
@protected
    MTObjectProp* _nested;
}

- (PrimitivePage*)nested { return ((MTPageRef*) _nested.value).page; }

- (NSArray*)props { return MT_PROPS(_nested, ); }

- (id)init {
    if ((self = [super init])) {
        _nested = [[MTObjectProp alloc] initWithPropSpec:s_nestedSpec];
    }
    return self;
}

+ (void)initialize {
    if (self == [RefPage class]) {
        s_nestedSpec = [[MTPropSpec alloc] initWithName:@"nested"
            annotations:nil
            valueClasses:@[ [MTPageRef class],[PrimitivePage class], ] ];
    }
}

@end
